cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java

```diff
@@ -122,15 +161,23 @@ public void setAuditPublisher(AuditPublisher auditPublisher) {
    * @throws NotFoundException if either the namespace or dataset instance is not found,
    * @throws IOException if there is a problem in making an HTTP request to check if the namespace exists.
    */
-  public DatasetMeta get(Id.DatasetInstance instance, List<? extends Id> owners) throws Exception {
+  public DatasetMeta get(final Id.DatasetInstance instance, List<? extends Id> owners) throws Exception {
+    // Application Deployment first makes a call to the dataset service to check if the instance already exists with
+    // a different type. To make sure that that call responds with the right exceptions if necessary, first fetch the
+    // meta from the cache and throw appropriate exceptions if necessary.
+    DatasetMeta datasetMeta;
     try {
-      return metaCache.get(instance);
+      datasetMeta = metaCache.get(instance);
     } catch (ExecutionException e) {
-      if ((e.getCause() instanceof Exception) && (e.getCause() instanceof HttpErrorStatusProvider)) {
-         throw (Exception) e.getCause();
+      Throwable cause = e.getCause();
+      if ((cause instanceof Exception) && (cause instanceof HttpErrorStatusProvider)) {
+         throw (Exception) cause;
       }
       throw e;
     }
+    // Only return the above datasetMeta if authorization succeeds
+    ensureAccess(instance.toEntityId());
```

Comment:
It is not clear what should be thrown... So the app attempts to create a dataset that it does not have access to? How can it even do that? If it has write access on the namespace then it has the right to create the dataset, correct? If the dataset already exists, and the app (user) does not have privilege to see that, what should be the correct exception? If it is a SecurityException, then it is obvious that the dataset exists, but there is not privilege. What what other exception would not convey that fact?